### PR TITLE
[FIX] account: Fixed inconsistent tax grids on journal entries

### PR DESCRIPTION
**Pull request by hsal-odoo**
This commit ensure that tax_tag_ids are cleared when tax_ids are cleared on the account_move_line.

Before this commit, when a new journal entry is created, adding a tax with tax grids then removing the tax does not remove the grids.

task-3906808

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by hsal-odoo** [comment]
@william-andre There might be a better solution to the problem, but I spent some time trying to figure out another solution but could not find any. The problem I could identify was that the `_compute_all_tax()` compute function does not trigger when `tax_ids` is emptied.

File: addons/account/models/account_move_line.py [hunk 1]
```diff
@@ -1185,6 +1185,12 @@ def action_register_payment(self):
     # INVERSE METHODS
     # -------------------------------------------------------------------------
 
+    @api.onchange('tax_ids')
```
Comment (william-andre) [inline addons/account/models/account_move_line.py:1188]: This only works in the UI but won't work if the tax is removed programmatically

Comment (hsal-odoo) [reply addons/account/models/account_move_line.py:1188]: and what is a better solution in that case?

Comment (hsal-odoo) [reply addons/account/models/account_move_line.py:1188]: should I override the write method?

Comment (william-andre) [reply addons/account/models/account_move_line.py:1188]: The issue is probably something in `_sync_dynamic_lines` that doesn't work

Comment (william-andre) [reply addons/account/models/account_move_line.py:1188]: To get started, let's start by writing a test :wink: (not using the `Form` emulator)

Comment (hsal-odoo) [reply addons/account/models/account_move_line.py:1193]: @william-andre I have wrote the unit test, plz have a look :)

**Review by william-andre — COMMENTED** [review]
@hsal-odoo I pushed a fixup commit, please review it.

Note that your fix didn't pass your test :wink:

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -2415,29 +2415,31 @@ def _apply_cash_rounding(self, diff_balance, diff_amount_currency, cash_rounding
 
     @contextmanager
     def _sync_unbalanced_lines(self, container):
+        def has_tax(move):
+            return bool(move.line_ids.tax_ids)
+
+        move_had_tax = {move: has_tax(move) for move in container['records']}
         yield
+        move_has_tax = {move: has_tax(move) for move in container['records']}
```
Comment (hsal-odoo) [inline addons/account/models/account_move.py:2423]: @william-andre This dict is not needed, `has_tax` can be directly called after the yield when needed

File: addons/account/models/account_move.py [hunk 3]
```diff
@@ -2415,29 +2415,31 @@ def _apply_cash_rounding(self, diff_balance, diff_amount_currency, cash_rounding
...
-
-            # Unlink tax lines if all taxes have been removed.
-            if not invoice.line_ids.tax_ids:
-                # if there isn't any tax but there remains a tax_line_id, it means we are currently in the process of
-                # removing the taxes from the entry. Thus, we want the automatic balancing to happen in order  to have
-                # a smooth process for tax deletion
-                if not invoice.line_ids.filtered('tax_line_id'):
-                    continue
-                invoice.line_ids.filtered('tax_line_id').unlink()
+        for move in (x for x in container['records'] if x.state != 'posted'):
```
Comment (hsal-odoo) [inline addons/account/models/account_move.py:2425]: leaving the loop variable as `invoice` instead of `move` would eliminate some unnecessary diff

**Review by hsal-odoo — COMMENTED** [review]
I think I grasp the logic in the commit, but I think it can be improved, and its commit message needs to be improved as well. Should I do these fixes myself?

File: addons/account/models/account_move.py (hunk 3 above)
Comment (william-andre) [reply addons/account/models/account_move.py:2425]: It is very confusing because it will never be an invoice.

**Review by william-andre — APPROVED** [review]
@robodoo r+
